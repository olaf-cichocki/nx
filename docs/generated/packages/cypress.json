{
  "githubRoot": "https://github.com/nrwl/nx/blob/master",
  "name": "cypress",
  "description": "The Nx Plugin for Cypress contains executors and generators allowing your workspace to use the powerful Cypress integration testing capabilities.",
  "root": "/packages/cypress",
  "source": "/packages/cypress/src",
  "generators": [
    {
      "name": "init",
      "factory": "./src/generators/init/init#cypressInitGenerator",
      "schema": {
        "$schema": "http://json-schema.org/schema",
        "$id": "NxCypressInit",
        "cli": "nx",
        "title": "Add Cypress Configuration to the workspace",
        "description": "Add Cypress Configuration to the workspace.",
        "type": "object",
        "properties": {
          "skipPackageJson": {
            "type": "boolean",
            "default": false,
            "description": "Do not add dependencies to `package.json`."
          }
        },
        "presets": []
      },
      "description": "Initialize the `@nrwl/cypress` plugin.",
      "aliases": ["ng-add"],
      "hidden": true,
      "implementation": "/packages/cypress/src/generators/init/init#cypressInitGenerator.ts",
      "path": "/packages/cypress/src/generators/init/schema.json"
    },
    {
      "name": "cypress-project",
      "factory": "./src/generators/cypress-project/cypress-project#cypressProjectGenerator",
      "schema": {
        "$schema": "http://json-schema.org/schema",
        "$id": "NxCypressProjectGeneratorSchema",
        "cli": "nx",
        "title": "Create Cypress Configuration for the workspace",
        "description": "Create Cypress Configuration for the workspace.",
        "type": "object",
        "properties": {
          "project": {
            "type": "string",
            "description": "The name of the frontend project to test.",
            "$default": { "$source": "projectName" }
          },
          "baseUrl": {
            "type": "string",
            "description": "The address (with the port) which your application is running on."
          },
          "name": {
            "type": "string",
            "description": "Name of the E2E Project.",
            "$default": { "$source": "argv", "index": 0 },
            "x-prompt": "What name would you like to use for the e2e project?"
          },
          "directory": {
            "type": "string",
            "description": "A directory where the project is placed."
          },
          "linter": {
            "description": "The tool to use for running lint checks.",
            "type": "string",
            "enum": ["eslint", "tslint", "none"],
            "default": "eslint"
          },
          "js": {
            "description": "Generate JavaScript files rather than TypeScript files.",
            "type": "boolean",
            "default": false
          },
          "skipFormat": {
            "description": "Skip formatting files.",
            "type": "boolean",
            "default": false
          },
          "setParserOptionsProject": {
            "type": "boolean",
            "description": "Whether or not to configure the ESLint `parserOptions.project` option. We do not do this by default for lint performance reasons.",
            "default": false
          },
          "standaloneConfig": {
            "description": "Split the project configuration into `<projectRoot>/project.json` rather than including it inside workspace.json.",
            "type": "boolean"
          },
          "skipPackageJson": {
            "type": "boolean",
            "default": false,
            "description": "Do not add dependencies to `package.json`."
          }
        },
        "required": ["name"],
        "presets": []
      },
      "description": "Add a Cypress E2E Project.",
      "hidden": true,
      "implementation": "/packages/cypress/src/generators/cypress-project/cypress-project#cypressProjectGenerator.ts",
      "aliases": [],
      "path": "/packages/cypress/src/generators/cypress-project/schema.json"
    },
    {
      "name": "cypress-component-project",
      "factory": "./src/generators/cypress-component-project/cypress-component-project#cypressComponentProject",
      "schema": {
        "$schema": "http://json-schema.org/schema",
        "$id": "NxCypressComponentProject",
        "cli": "nx",
        "title": "Set up Cypress component test for a project",
        "type": "object",
        "examples": [
          {
            "command": "nx g @nrwl/cypress:cypress-component-project --project=my-cool-lib --componentType=react",
            "description": "Add cypress component testing to an existing project named my-cool-lib"
          },
          {
            "command": "nx g @nrwl/cypress:cypress-component-project --project=my-cool-lib --componentType=react --force",
            "description": "Add Cypress component testing to an existing project that already has a Cypress component testing enabled. Helpful to reset the configuration to the default."
          }
        ],
        "properties": {
          "project": {
            "type": "string",
            "description": "The name of the project to convert",
            "$default": { "$source": "projectName" },
            "x-prompt": "What project should we add Cypress component testing to?"
          },
          "componentType": {
            "type": "string",
            "enum": ["react", "next"],
            "description": "Which framework is being used?",
            "alias": "framework",
            "x-prompt": "What framework is being used in this project?"
          },
          "compiler": {
            "type": "string",
            "enum": ["swc", "babel"],
            "description": "Which compiler is being used in the existing project?",
            "default": "babel"
          },
          "force": {
            "type": "boolean",
            "description": "Overwrite an existing cypress component test project. This is helpful if you want to reset to the default configurations.",
            "default": false
          }
        },
        "required": ["project", "componentType"],
        "presets": []
      },
      "description": "Set up Cypress Component Test for a project",
      "aliases": ["cy-cmp"],
      "hidden": true,
      "implementation": "/packages/cypress/src/generators/cypress-component-project/cypress-component-project#cypressComponentProject.ts",
      "path": "/packages/cypress/src/generators/cypress-component-project/schema.json"
    },
    {
      "name": "convert-cypress-ten",
      "factory": "./src/generators/convert-to-cypress-10/convert-to-cypress-ten#convertCypressProject",
      "schema": {
        "$schema": "http://json-schema.org/schema",
        "$id": "NxCypressConvertToTen",
        "cli": "nx",
        "title": "Convert e2e project to Cypress 10",
        "type": "object",
        "examples": [
          {
            "command": "nx g @nrwl/cypress:convert-cypress-ten --project=my-cool-app-e2e",
            "description": "Convert my-cool-app-e2e to Cypress 10"
          },
          {
            "command": "nx g @nrwl/cypress:convert-cypress-ten --all",
            "description": "Convert all e2e projects in the workspace to Cypress 10"
          }
        ],
        "properties": {
          "project": {
            "type": "string",
            "description": "The name of the project to convert",
            "$default": { "$source": "projectName" }
          },
          "all": {
            "type": "boolean",
            "description": "Convert all projects in the workspace",
            "default": false
          },
          "targets": {
            "type": "array",
            "description": "The targets to convert",
            "default": ["e2e"]
          }
        },
        "presets": []
      },
      "aliases": ["cy10"],
      "description": "Convert to Cypress 10",
      "hidden": true,
      "implementation": "/packages/cypress/src/generators/convert-to-cypress-10/convert-to-cypress-ten#convertCypressProject.ts",
      "path": "/packages/cypress/src/generators/convert-to-cypress-10/schema.json"
    },
    {
      "name": "cypress-component-test-file",
      "factory": "./src/generators/cypress-component-test-file/cypress-component-test-file#cypressComponentTestFiles",
      "schema": {
        "$schema": "http://json-schema.org/schema",
        "$id": "NxCypressComponentTestFileProject",
        "cli": "nx",
        "title": "Add Cypress Component test file to project",
        "type": "object",
        "examples": [],
        "properties": {
          "project": {
            "type": "string",
            "description": "The name of the project",
            "$default": { "$source": "projectName" },
            "x-prompt": "What is the name of the project for this component test file?"
          },
          "name": {
            "type": "string",
            "description": "The name of the component to be tested",
            "$default": { "$source": "argv", "index": 0 },
            "x-prompt": "What name would you like to use for the component test file?"
          },
          "componentType": {
            "type": "string",
            "enum": ["react", "next"],
            "description": "Which framework is being used?",
            "alias": "framework",
            "x-prompt": "What framework is being used in this project?"
          },
          "directory": {
            "type": "string",
            "description": "The directory where the component test file should be created from the projects source root",
            "alias": "dir"
          }
        },
        "required": ["name", "project", "componentType"],
        "presets": []
      },
      "alias": ["cy-test"],
      "description": "Create a Cypress component test for a given component",
      "hidden": true,
      "implementation": "/packages/cypress/src/generators/cypress-component-test-file/cypress-component-test-file#cypressComponentTestFiles.ts",
      "aliases": [],
      "path": "/packages/cypress/src/generators/cypress-component-test-file/schema.json"
    }
  ],
  "executors": [
    {
      "name": "cypress",
      "implementation": "/packages/cypress/src/executors/cypress/cypress.impl.ts",
      "schema": {
        "title": "Cypress Target",
        "description": "Cypress target option for Build Facade.",
        "type": "object",
        "outputCapture": "pipe",
        "cli": "nx",
        "properties": {
          "cypressConfig": {
            "type": "string",
            "description": "The path of the Cypress configuration json file."
          },
          "watch": {
            "type": "boolean",
            "description": "Recompile and run tests when files change.",
            "default": false
          },
          "tsConfig": {
            "type": "string",
            "description": "The path of the Cypress tsconfig configuration json file."
          },
          "devServerTarget": {
            "type": "string",
            "description": "Dev server target to run tests against."
          },
          "headed": {
            "type": "boolean",
            "description": "Displays the browser instead of running headlessly. Set this to `true` if your run depends on a Chrome extension being loaded.",
            "default": false
          },
          "headless": {
            "type": "boolean",
            "description": "Hide the browser instead of running headed (default for cypress run).",
            "default": false,
            "x-deprecated": true
          },
          "exit": {
            "type": "boolean",
            "description": "Whether or not the Cypress Test Runner will stay open after running tests in a spec file.",
            "default": true
          },
          "key": {
            "type": "string",
            "description": "The key cypress should use to run tests in parallel/record the run (CI only)."
          },
          "record": {
            "type": "boolean",
            "description": "Whether or not Cypress should record the results of the tests.",
            "default": false
          },
          "parallel": {
            "type": "boolean",
            "description": "Whether or not Cypress should run its tests in parallel (CI only).",
            "default": false
          },
          "baseUrl": {
            "type": "string",
            "description": "The address (with the port) which your application is running on."
          },
          "browser": {
            "type": "string",
            "description": "The browser to run tests in."
          },
          "env": {
            "type": "object",
            "description": "A key-value Pair of environment variables to pass to Cypress runner."
          },
          "spec": {
            "type": "string",
            "description": "A comma delimited glob string that is provided to the Cypress runner to specify which spec files to run. i.e. `**examples/**,**actions.spec**`."
          },
          "copyFiles": {
            "type": "string",
            "description": "A regex string that is used to choose what additional integration files to copy to the dist folder.",
            "x-deprecated": true
          },
          "ciBuildId": {
            "oneOf": [{ "type": "string" }, { "type": "number" }],
            "description": "A unique identifier for a run to enable grouping or parallelization."
          },
          "group": {
            "type": "string",
            "description": "A named group for recorded runs in the Cypress dashboard."
          },
          "ignoreTestFiles": {
            "type": "string",
            "description": "A String or Array of glob patterns used to ignore test files that would otherwise be shown in your list of tests. Cypress uses minimatch with the options: `{dot: true, matchBase: true}`. We suggest using https://globster.xyz to test what files would match."
          },
          "reporter": {
            "type": "string",
            "description": "The reporter used during cypress run."
          },
          "reporterOptions": {
            "type": "string",
            "description": "The reporter options used. Supported options depend on the reporter."
          },
          "skipServe": {
            "type": "boolean",
            "description": "Skip dev-server build.",
            "default": false
          },
          "testingType": {
            "type": "string",
            "description": "Specify the type of tests to execute.",
            "enum": ["component", "e2e"],
            "default": "e2e"
          },
          "tag": {
            "type": "string",
            "description": "A comma delimited list to identify a run with.",
            "aliases": ["t"]
          }
        },
        "additionalProperties": true,
        "required": ["cypressConfig"],
        "presets": []
      },
      "hasher": "./src/executors/cypress/hasher",
      "description": "Run Cypress E2E tests.",
      "aliases": [],
      "hidden": false,
      "path": "/packages/cypress/src/executors/cypress/schema.json"
    }
  ]
}
