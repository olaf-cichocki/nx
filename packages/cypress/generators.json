{
  "name": "Nx Cypress",
  "version": "0.1",
  "schematics": {
    "init": {
      "factory": "./src/generators/init/init#cypressInitSchematic",
      "schema": "./src/generators/init/schema.json",
      "description": "Initialize the `@nrwl/cypress` plugin.",
      "aliases": ["ng-add"],
      "hidden": true
    },
    "cypress-project": {
      "factory": "./src/generators/cypress-project/cypress-project#cypressProjectSchematic",
      "schema": "./src/generators/cypress-project/schema.json",
      "description": "Add a Cypress E2E Project"
    },
    "cypress-component-project": {
      "factory": "./src/generators/cypress-component-project/cypress-component-project#cypressComponentProject",
      "schema": "./src/generators/cypress-component-project/schema.json",
      "aliases": ["cy-cmp"],
      "description": "Set up Cypress Component Test for a project"
    },
    "cypress-component-test-file": {
      "factory": "./src/generators/cypress-component-test-file/cypress-component-test-file#cypressComponentTestFiles",
      "schema": "./src/generators/cypress-component-test-file/schema.json",
      "alias": ["cy-test"],
      "description": "Create a Cypress component test for a given component"
    },
    "convert-cypress-ten": {
      "factory": "./src/generators/convert-to-cypress-10/convert-to-cypress-ten#convertCypressProject",
      "schema": "./src/generators/convert-to-cypress-10/schema.json",
      "aliases": ["cy10"],
      "description": "Convert existing Cypress e2e projects to Cypress 10"
    }
  },
  "generators": {
    "init": {
      "factory": "./src/generators/init/init#cypressInitGenerator",
      "schema": "./src/generators/init/schema.json",
      "description": "Initialize the `@nrwl/cypress` plugin.",
      "aliases": ["ng-add"],
      "hidden": true
    },
    "cypress-project": {
      "factory": "./src/generators/cypress-project/cypress-project#cypressProjectGenerator",
      "schema": "./src/generators/cypress-project/schema.json",
      "description": "Add a Cypress E2E Project.",
      "hidden": true
    },
    "cypress-component-project": {
      "factory": "./src/generators/cypress-component-project/cypress-component-project#cypressComponentProject",
      "schema": "./src/generators/cypress-component-project/schema.json",
      "description": "Set up Cypress Component Test for a project",
      "aliases": ["cy-cmp"],
      "hidden": true
    },
    "convert-cypress-ten": {
      "factory": "./src/generators/convert-to-cypress-10/convert-to-cypress-ten#convertCypressProject",
      "schema": "./src/generators/convert-to-cypress-10/schema.json",
      "aliases": ["cy10"],
      "description": "Convert to Cypress 10",
      "hidden": true
    },
    "cypress-component-test-file": {
      "factory": "./src/generators/cypress-component-test-file/cypress-component-test-file#cypressComponentTestFiles",
      "schema": "./src/generators/cypress-component-test-file/schema.json",
      "alias": ["cy-test"],
      "description": "Create a Cypress component test for a given component",
      "hidden": true
    }
  }
}
