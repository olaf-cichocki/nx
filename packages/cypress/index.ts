export { cypressProjectGenerator } from './src/generators/cypress-project/cypress-project';
export { cypressInitGenerator } from './src/generators/init/init';
export { conversionGenerator } from './src/generators/convert-tslint-to-eslint/convert-tslint-to-eslint';
export { cypressComponentProject } from './src/generators/cypress-component-project/cypress-component-project';
export { convertCypressProject } from './src/generators/convert-to-cypress-10/convert-to-cypress-ten';
export { CYPRESS_COMPONENT_TEST_TARGET } from './src/utils/project-name';
export { cypressComponentTestFiles } from './src/generators/cypress-component-test-file/cypress-component-test-file';
