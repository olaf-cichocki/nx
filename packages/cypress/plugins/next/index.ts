import { workspaceLayout } from 'nx/src/core/file-utils';
import { workspaceRoot } from 'nx/src/utils/app-root';
import { join, resolve } from 'path';
// doesn't like ESM imports
const findNextWebpackConfig = require('@cypress/react/plugins/next/findNextWebpackConfig');

/**
 * Start the devserver for Cypress component tests with a NextJS project.
 */
export function componentDevServer() {
  return async (cypressConfigOptions) => {
    // when getting the next config it will look for a next.config.js file and run it.
    // this will run the withNx() plugin which adds support for css modules and libs.
    // also allows for extending the webpack config as needed.
    const webpackConfig = await findNextWebpackConfig(
      cypressConfigOptions.config
    );

    if (!webpackConfig) {
      throw new Error(
        'Could not find a webpack config for next. is @nrwl/next installed?'
      );
    }
    const nextLoaders = webpackConfig.module.rules.find(
      (rule: any) => typeof rule.oneOf === 'object'
    );

    const moduleLoaders = nextLoaders.oneOf.filter((rule) =>
      regexEqual(rule.test, /\.(tsx|ts|js|cjs|mjs|jsx)$/)
    );

    if (moduleLoaders) {
      const { libsDir } = workspaceLayout();
      const libPath = join(workspaceRoot, libsDir);

      for (const moduleLoader of moduleLoaders) {
        moduleLoader.include.push(libPath);
      }
    }

    // needs to be imported after loading the webpack config from next;
    const { startDevServer } = require('@cypress/webpack-dev-server');

    return startDevServer({
      options: cypressConfigOptions,
      webpackConfig,
      template: resolve(__dirname, 'next.template.html'),
    });
  };
}

function regexEqual(x, y) {
  return (
    x instanceof RegExp &&
    y instanceof RegExp &&
    x.source === y.source &&
    x.global === y.global &&
    x.ignoreCase === y.ignoreCase &&
    x.multiline === y.multiline
  );
}
