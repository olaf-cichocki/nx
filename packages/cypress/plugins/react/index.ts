import { startDevServer } from '@cypress/webpack-dev-server';
import { Configuration } from 'webpack';
import { buildBaseWebpackConfig } from '../utils/webpack';

export function componentDevServer(
  tsConfigPath = 'tsconfig.cy.json',
  compiler: 'swc' | 'babel' = 'babel',
  extendWebpackConfig?: (config: Configuration) => Configuration
): (
  cypressDevServerConfig,
  devServerConfig
) => ReturnType<typeof startDevServer> {
  const NODE_ENV = 'test';

  if (!process.env.BABEL_ENV) {
    process.env.BABEL_ENV = NODE_ENV;
  }

  if (!process.env.NODE_ENV) {
    process.env.NODE_ENV = NODE_ENV;
  }

  let webpackConfig = buildBaseWebpackConfig({
    tsConfigPath,
    compiler,
    preset: 'react',
  });

  if (extendWebpackConfig) {
    webpackConfig = extendWebpackConfig(webpackConfig);
  }

  return (cypressDevServerConfig, devServerConfig) =>
    startDevServer({
      options: cypressDevServerConfig,
      webpackConfig: webpackConfig as never,
    });
}
