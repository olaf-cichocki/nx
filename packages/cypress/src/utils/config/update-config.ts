import {
  createPrinter,
  createSourceFile,
  ScriptTarget,
  transform,
} from 'typescript';
import {
  CypressConfigWithoutDevServer,
  mergeCypressConfigs,
} from './transformers';

// TODO(caleb): handle devserver in config
export function addOrUpdateConfigProperties(
  configContent: string,
  config: CypressConfigWithoutDevServer,
  overwrite: boolean = false
) {
  const sourceFile = createSourceFile(
    'cypress.config.ts',
    configContent,
    ScriptTarget.Latest,
    true
  );

  const transformedResult = transform(sourceFile, [
    mergeCypressConfigs(config, overwrite),
  ]);

  return createPrinter().printFile(transformedResult.transformed[0]);
}

export function removeConfigProperties() {}

export function addConfigImport() {}

export function removeConfigImport() {}
