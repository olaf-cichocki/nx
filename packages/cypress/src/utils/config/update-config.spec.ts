import { addOrUpdateConfigProperties } from './update-config';

describe('Update Cypress Config', () => {
  let configContent = `
import { defineConfig } from 'cypress';
import { componentDevServer } from '@nrwl/cypress/plugins/next';


export default defineConfig({
  component: {
    devServer: componentDevServer('tsconfig.cy.json', 'babel'),
    pluginsFile: false,
    video: true,
    chromeWebSecurity: false,
    fixturesFolder: 'cypress/fixtures',
    specPattern: '**/*.cy.{js,jsx,ts,tsx}',
    supportFile: 'cypress/support/component.ts',
    videosFolder: '../../dist/cypress/apps/n/videos',
    screenshotsFolder: '../../dist/cypress/apps/n/screenshots',
  },
  e2e: {
    fileServerFolder: '.',
    fixturesFolder: './src/fixtures',
    integrationFolder: './src/e2e',
    supportFile: './src/support/e2e.ts',
    specPattern: '**/*.cy.{js,ts}',
    pluginsFile: false,
    video: true,
    videosFolder: '../../dist/cypress/apps/myapp4299814-e2e/videos',
    screenshotsFolder: '../../dist/cypress/apps/myapp4299814-e2e/screenshots',
    chromeWebSecurity: false,
  },
});
  `;

  describe('Imports', () => {
    it('should add an import', () => {});

    it('should remove an import', () => {});

    it('should update an import', () => {});
  });

  describe('Properties', () => {
    it('should add a property', () => {
      const actual = addOrUpdateConfigProperties(configContent, {
        baseUrl: 'http://localhost:8080',
      });

      expect(actual).toMatchSnapshot();
    });

    it('should update a property', () => {});

    it('should update a nested property', () => {});

    it('should remove a property', () => {});

    it('should remove a nested property', () => {});
  });
});
