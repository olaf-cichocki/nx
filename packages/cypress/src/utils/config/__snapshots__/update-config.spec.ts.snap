// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Cypress Config Properties should add a property 1`] = `
"import { defineConfig } from 'cypress';
import { componentDevServer } from '@nrwl/cypress/plugins/next';
export default defineConfig({
    baseUrl: 'http://localhost:8080',
    component: {
        devServer: componentDevServer('tsconfig.cy.json', 'babel'),
        pluginsFile: false,
        video: true,
        chromeWebSecurity: false,
        fixturesFolder: 'cypress/fixtures',
        specPattern: '**/*.cy.{js,jsx,ts,tsx}',
        supportFile: 'cypress/support/component.ts',
        videosFolder: ''../../dist/cypress/apps/n/videos',
        screenshotsFolder: '../../dist/cypress/apps/n/screenshots',
    },
    e2e: {
        fileServerFolder: '.',
        fixturesFolder: './src/fixtures',
        integrationFolder: './src/e2e',
        supportFile: './src/support/e2e.ts',
        specPattern: '**/*.cy.{js,ts}',
        pluginsFile: false,
        video: true,
        videosFolder: '../../dist/cypress/apps/myapp4299814-e2e/videos',
        screenshotsFolder: '../../dist/cypress/apps/myapp4299814-e2e/screenshots',
        chromeWebSecurity: false,
    }
});
"
`;
